### PR TITLE
Add models and schema for backend objects and Results

### PR DESCRIPTION
**Pull request by diego-plan9**
### Summary

Follow-up to #1249, adding models and schemas for:
* BackendConfiguration
* BackendProperties
* BackendStatus
* JobStatus
* Results

along with the utilities needed for supporting their functionality (Fields, Validators, and tweaks and fixes for the base models).The models however have been tested against the existing `qiskit/schemas/examples` files by modifying the `test_schema.py` file.

By having these files in place, the rest of the work (updating the components needed for each schema, adjusting the simulators for conforming to it, working with results, etc) can hopefully be split more easily.

### Details and comments

Please note that the objects are not in use yet - will be tackled in subsequent PRS. Also, the location of some files and classes is still not set in stone (in particular, `qiskit.validation.result` will be removed, Jay!). No changes have been made to the schemas themselves, and the .json files have been used as a reference.

The schema files

### CODE REVIEW

File: qiskit/validation/result.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+    counts = Nested(ObjSchema,
+                    validate=PatternProperties(
+                        {Regexp('^0x([0-9A-Fa-f])+$'): Integer()}))
+    snapshots = Nested(ObjSchema)
+    memory = List(Raw(),
+                  validate=Length(min=1))
+    statevector = List(List(Complex(),
+                            validate=Length(min=1)))
+    unitary = List(List(List(Complex(),
```
Comment (delapuente) [inline qiskit/validation/result.py]: Given the level of nesting of the fields, I think we could increase legibility by intending field definitions as:
```python
unitary = List(
  List(
    List(Complex, validate=Length(min=1)),
    validate=Length(min=1)
  ),
  validate=Length(min=1)
)
```

File: qiskit/validation/result.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        shots (int or tuple): the starting and ending shot for this data.
+        success (bool): if true, we can trust results for this experiment.
+        data (ExperimentResultData): results information.
+    """
+
+    def __init__(self, shots, success, data, **kwargs):
+        self.shots = shots
+        self.success = success
+        self.data = data
+        kwargs.update(self.__dict__)
```
Comment (delapuente) [inline qiskit/validation/result.py]: Would you mind to clarify why is this line needed?

File: qiskit/validation/validate.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        for validator in self.validators:
+            try:
+                return validator(value)
+            except ValidationError:
+                pass
+
+        raise ValidationError('nope')
+
+
+class PatternProperties(Validator):
```
Comment (delapuente) [inline qiskit/validation/validate.py]: This validator seems to me like a kind of schema where some fields can be added dynamically if they follow some patterns. Also, notice that it is partially repeating what a schema does by trying to deserialize the values of each property which is the main job of a schema.

As mentioned before, since the usage of this is confined to modelling `counts`, I'd recommend to use it internally in the suggested alternative model of `counts`.

File: qiskit/validation/result.py (hunk 2 above)
Comment (diego-plan9) [reply qiskit/validation/result.py]: Good catch - it's a leftover from previous iterations, where the validation of the `BaseModel`s was done _after_ instantiation. This in turn required `**kwargs` (for the super call) to contain all the arguments, and it was a quick and not verbose way of updating `kwargs) in order to contain them. I'll remove them as they are no longer needed.

**Comment by diego-plan9** [comment]
Thanks for the review, @delapuente !

Regarding the `counts` field, the motivation was mostly to explicitly avoid using `Dict` (the field), as previous exploratory efforts yielded some errors (that might be no longer a concern, but part of using it is ... testing it as well). Consistency also played a role - it would be the only item in the current schemas that would be a `Dict`, and it felt safer to stick to the overall approach and model it in a similar way (with the extra validator) - but indeed is a blurry line.

Combined with the nature of the PR (setting the basis for future PRs, and allow for adjustments of different kinds once the rest of the work is completed), I went with the "conservative" approach - I think your suggestion is bound to be explored, although perhaps it fits better to make it as a follow-up for optimizing efforts.
